File: compensation/compensation-domain/src/main/kotlin/me/ahoo/wow/compensation/domain/ExecutionFailedState.kt
/*
 * Copyright [2021-present] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.ahoo.wow.compensation.domain

import me.ahoo.wow.api.annotation.OnSourcing
import me.ahoo.wow.api.messaging.FunctionKind
import me.ahoo.wow.api.messaging.processor.ProcessorInfoData
import me.ahoo.wow.compensation.api.CompensationPrepared
import me.ahoo.wow.compensation.api.ErrorDetails
import me.ahoo.wow.compensation.api.EventId
import me.ahoo.wow.compensation.api.ExecutionFailedApplied
import me.ahoo.wow.compensation.api.ExecutionFailedCreated
import me.ahoo.wow.compensation.api.ExecutionFailedStatus
import me.ahoo.wow.compensation.api.ExecutionSuccessApplied
import me.ahoo.wow.compensation.api.IExecutionFailedState

class ExecutionFailedState(override val id: String) : IExecutionFailedState {
    override lateinit var eventId: EventId
        private set
    override lateinit var processor: ProcessorInfoData
        private set
    override lateinit var functionKind: FunctionKind
        private set
    override lateinit var error: ErrorDetails
        private set
    override var executionTime: Long = 0
        private set
    override var status: ExecutionFailedStatus = ExecutionFailedStatus.FAILED
        private set
    override var retriedTimes: Int = 0
        private set

    @OnSourcing
    fun onCreated(event: ExecutionFailedCreated) {
        this.eventId = event.eventId
        this.processor = event.processor
        this.functionKind = event.functionKind
        this.error = event.error
        this.executionTime = event.executionTime
        this.status = ExecutionFailedStatus.FAILED
    }

    @OnSourcing
    fun onPrepared(event: CompensationPrepared) {
        this.status = ExecutionFailedStatus.PREPARED
    }

    @OnSourcing
    fun onFailed(event: ExecutionFailedApplied) {
        this.retriedTimes = event.retriedTimes
        this.executionTime = event.executionTime
        this.status = ExecutionFailedStatus.FAILED
    }

    @OnSourcing
    fun onSuccess(event: ExecutionSuccessApplied) {
        this.retriedTimes = event.retriedTimes
        this.executionTime = event.executionTime
        this.status = ExecutionFailedStatus.SUCCEEDED
    }
}

File: compensation/wow-compensation-dashboard/src/app/api/ExecutionFailedState.ts
/*
 * Copyright [2021-present] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * compensation.IExecutionFailedState
 */
export interface ExecutionFailedState {
  id: string;
  eventId: EventId;
  error: ErrorDetails;
  executeAt: number;
  functionKind: FunctionKind;
  isRetryable: boolean;
  processor: ProcessorInfoData;
  retrySpec: RetrySpec;
  retryState: RetryState;
  status: ExecutionFailedStatus;
  recoverable: RecoverableType;
}

/**
 * compensation.ErrorDetails
 */
export interface ErrorDetails {
  errorCode: string;
  errorMsg: string;
  stackTrace: string;
  succeeded: boolean;
}

/**
 * compensation.EventId
 */
export interface EventId {
  aggregateId: AggregateId;
  id: string;
  version: number;
}

/**
 * wow.AggregateId
 */
export interface AggregateId {
  aggregateId: string;
  aggregateName: string;
  contextName: string;
  tenantId: string;
}

export enum FunctionKind {
  Command = "COMMAND",
  Error = "ERROR",
  Event = "EVENT",
  Sourcing = "SOURCING",
  StateEvent = "STATE_EVENT",
}

/**
 * wow.ProcessorInfoData
 */
export interface ProcessorInfoData {
  contextName: string;
  processorName: string;
}

/**
 * compensation.RetrySpec
 */
export interface RetrySpec {
  executionTimeout: number;
  maxRetries: number;
  minBackoff: number;
}

/**
 * compensation.RetryState
 */
export interface RetryState {
  nextRetryAt: number;
  retries: number;
  retryAt: number;
  timeoutAt: number;
}

export enum ExecutionFailedStatus {
  FAILED = "FAILED",
  PREPARED = "PREPARED",
  SUCCEEDED = "SUCCEEDED",
}

export enum RecoverableType {
  /**
   * The exception is recoverable.
   */
  RECOVERABLE = "RECOVERABLE",

  /**
   * The exception is not recoverable.
   */
  UNRECOVERABLE = "UNRECOVERABLE",

  /**
   * The exception is unknown.
   */
  UNKNOWN = "UNKNOWN"
}
